File: backend/src/services/docker.service.ts
import Docker, { type ContainerInfo, type ContainerInspectInfo, type ContainerStats as DockerContainerStats } from 'dockerode';
import { getDockerSocketPath } from '../utils/dockerSocket';
import { logger } from '../utils/logger';
import type { Container, ContainerDetails, ContainerStats, LogOptions } from '../types/container.types';

export class DockerService {
  private docker: Docker;

  constructor() {
    const socketPath = getDockerSocketPath();
    this.docker = new Docker({ socketPath });
    logger.info(`Docker service initialized with socket: ${socketPath}`);
  }

  async listContainers(all = false): Promise<Container[]> {
    try {
      const containers = await this.docker.listContainers({ all });
      return containers.map(this.transformContainer);
    } catch (error) {
      logger.error('Failed to list containers:', error);
      throw new Error('Failed to list containers');
    }
  }

  async getContainer(id: string): Promise<ContainerDetails> {
    try {
      const container = this.docker.getContainer(id);
      const details = await container.inspect();
      return this.transformContainerDetails(details);
    } catch (error) {
      logger.error(`Failed to get container ${id}:`, error);
      throw new Error('Container not found');
    }
  }

  async startContainer(id: string): Promise<void> {
    try {
      const container = this.docker.getContainer(id);
      await container.start();
      logger.info(`Started container ${id}`);
    } catch (error) {
      logger.error(`Failed to start container ${id}:`, error);
      throw new Error('Failed to start container');
    }
  }

  async stopContainer(id: string): Promise<void> {
    try {
      const container = this.docker.getContainer(id);
      await container.stop({ t: 10 });
      logger.info(`Stopped container ${id}`);
    } catch (error) {
      logger.error(`Failed to stop container ${id}:`, error);
      throw new Error('Failed to stop container');
    }
  }

  async restartContainer(id: string): Promise<void> {
    try {
      const container = this.docker.getContainer(id);
      await container.restart({ t: 10 });
      logger.info(`Restarted container ${id}`);
    } catch (error) {
      logger.error(`Failed to restart container ${id}:`, error);
      throw new Error('Failed to restart container');
    }
  }

  async pauseContainer(id: string): Promise<void> {
    try {
      const container = this.docker.getContainer(id);
      await container.pause();
      logger.info(`Paused container ${id}`);
    } catch (error) {
      logger.error(`Failed to pause container ${id}:`, error);
      throw new Error('Failed to pause container');
    }
  }

  async unpauseContainer(id: string): Promise<void> {
    try {
      const container = this.docker.getContainer(id);
      await container.unpause();
      logger.info(`Unpaused container ${id}`);
    } catch (error) {
      logger.error(`Failed to unpause container ${id}:`, error);
      throw new Error('Failed to unpause container');
    }
  }

  async removeContainer(id: string, force = false): Promise<void> {
    try {
      const container = this.docker.getContainer(id);
      await container.remove({ force, v: true });
      logger.info(`Removed container ${id}`);
    } catch (error) {
      logger.error(`Failed to remove container ${id}:`, error);
      throw new Error('Failed to remove container');
    }
  }

  async getContainerStats(id: string): Promise<ContainerStats> {
    try {
      const container = this.docker.getContainer(id);
      const stats = await container.stats({ stream: false });
      return this.transformStats(stats, id);
    } catch (error) {
      logger.error(`Failed to get stats for container ${id}:`, error);
      throw new Error('Failed to get container stats');
    }
  }

  async getContainerLogs(
    id: string,
    options: LogOptions = {}
  ): Promise<any> {
    try {
      const container = this.docker.getContainer(id);
      return await container.logs(options as any);
    } catch (error) {
      logger.error(`Failed to get logs for container ${id}:`, error);
      throw new Error('Failed to get container logs');
    }
  }

  async execInContainer(
    id: string,
    cmd: string[],
    env: Record<string, string> = {}
  ): Promise<{ exitCode: number; output: string }> {
    try {
      const container = this.docker.getContainer(id);
      const exec = await container.exec({
        Cmd: cmd,
        AttachStdout: true,
        AttachStderr: true,
        Env: Object.entries(env).map(([k, v]) => `${k}=${v}`),
      });

      const stream = await exec.start({ Detach: false });

      return new Promise((resolve, reject) => {
        let output = '';
        stream.on('data', (chunk: Buffer) => {
          output += chunk.toString();
        });

        stream.on('end', async () => {
          const info = await exec.inspect();
          resolve({ exitCode: info.ExitCode || 0, output });
        });

        stream.on('error', reject);
      });
    } catch (error) {
      logger.error(`Failed to exec in container ${id}:`, error);
      throw new Error('Failed to execute command in container');
    }
  }

  private transformContainer(dockerContainer: ContainerInfo): Container {
    return {
      id: dockerContainer.Id,
      names: dockerContainer.Names.map((n: string) => n.replace(/^\//, '')),
      image: dockerContainer.Image,
      imageId: dockerContainer.ImageID,
      command: dockerContainer.Command,
      created: dockerContainer.Created,
      state: dockerContainer.State,
      status: dockerContainer.Status,
      ports: (dockerContainer.Ports || []).map((p: any) => ({
        IP: p.IP,
        privatePort: p.PrivatePort,
        publicPort: p.PublicPort,
        type: p.Type as 'tcp' | 'udp',
      })),
      labels: dockerContainer.Labels || {},
    };
  }

  private transformContainerDetails(dockerDetails: ContainerInspectInfo): ContainerDetails {
    const container = this.transformContainer({
      Id: dockerDetails.Id,
      Names: dockerDetails.Name ? [dockerDetails.Name.replace(/^\//, '')] : [],
      Image: dockerDetails.Config?.Image || '',
      ImageID: dockerDetails.Image || '',
      Command: dockerDetails.Config?.Cmd?.join(' ') || '',
      Created: Math.floor(new Date(dockerDetails.Created).getTime() / 1000),
      State: dockerDetails.State?.Status || '',
      Status: dockerDetails.State?.Status || '',
      Ports: [],
      Labels: dockerDetails.Config?.Labels || {},
      SizeRw: 0,
      SizeRootFs: 0,
      HostConfig: { NetworkMode: '' },
      NetworkSettings: { Networks: {} },
      Mounts: [],
    } as any);

    return {
      ...container,
      hostConfig: {
        portBindings: dockerDetails.HostConfig?.PortBindings as any,
        binds: dockerDetails.HostConfig?.Binds,
        restartPolicy: dockerDetails.HostConfig?.RestartPolicy as any,
      },
      config: {
        labels: dockerDetails.Config?.Labels,
        env: dockerDetails.Config?.Env,
        cmd: dockerDetails.Config?.Cmd,
        entrypoint: dockerDetails.Config?.Entrypoint as any,
        workingDir: dockerDetails.Config?.WorkingDir,
        user: dockerDetails.Config?.User,
      },
      networkSettings: {
        networks: dockerDetails.NetworkSettings?.Networks as any,
        ipAddress: (dockerDetails.NetworkSettings as any)?.IPAddress,
        ipPrefixLen: (dockerDetails.NetworkSettings as any)?.IPPrefixLen,
        gateway: (dockerDetails.NetworkSettings as any)?.Gateway,
      },
      mounts: (dockerDetails.Mounts || []).map((m: any) => ({
        type: m.Type as 'bind' | 'volume' | 'tmpfs',
        source: m.Source,
        destination: m.Destination,
        mode: m.Mode,
        rw: m.RW,
        propagation: m.Propagation,
      })),
    };
  }

  private transformStats(dockerStats: DockerContainerStats, id: string): ContainerStats {
    const cpuDelta = dockerStats.cpu_stats.cpu_usage.total_usage - dockerStats.precpu_stats.cpu_usage.total_usage;
    const systemDelta = dockerStats.cpu_stats.system_cpu_usage - dockerStats.precpu_stats.system_cpu_usage;
    const cpuPercent = systemDelta > 0 ? (cpuDelta / systemDelta) * 100 * dockerStats.cpu_stats.online_cpus : 0;

    const memoryUsage = (dockerStats.memory_stats.usage as number) || 0;
    const memoryLimit = (dockerStats.memory_stats.limit as number) || 0;
    const memoryPercent = memoryLimit > 0 ? (memoryUsage / memoryLimit) * 100 : 0;

    const networks = Object.values(dockerStats.networks || {}) as any[];
    const network = networks.reduce(
      (acc: { rx: number; tx: number }, n: any) => ({
        rx: acc.rx + (n.rx_bytes || 0),
        tx: acc.tx + (n.tx_bytes || 0),
      }),
      { rx: 0, tx: 0 }
    );

    const ioStats = dockerStats.blkio_stats?.io_service_bytes_recursive || [];
    const blockRead = ioStats
      .filter((io: any) => io.op === 'Read')
      .reduce((sum: number, io: any) => sum + (io.value || 0), 0);
    const blockWrite = ioStats
      .filter((io: any) => io.op === 'Write')
      .reduce((sum: number, io: any) => sum + (io.value || 0), 0);

    return {
      name: dockerStats.name || '',
      id,
      cpuPercent: Math.round(cpuPercent * 100) / 100,
      memoryUsage,
      memoryLimit,
      memoryPercent: Math.round(memoryPercent * 100) / 100,
      netRx: network.rx,
      netTx: network.tx,
      blockRead,
      blockWrite,
    };
  }
}

export const dockerService = new DockerService();

File: backend/app/services/docker_service.py
from __future__ import annotations

from collections.abc import Iterable
from typing import Any

import docker

from app.config import config
from app.utils.logger import logger


class DockerService:
    def __init__(self) -> None:
        base_url = self._build_base_url(config.docker_socket_path)
        self.client = docker.DockerClient(base_url=base_url)
        self.api = docker.APIClient(base_url=base_url)
        logger.info("Docker service initialized with socket %s", config.docker_socket_path)

    @staticmethod
    def _build_base_url(socket_path: str) -> str:
        if socket_path.startswith("unix://") or socket_path.startswith("tcp://"):
            return socket_path
        return f"unix://{socket_path}"

    def list_containers(self, all_containers: bool = False) -> list[dict[str, Any]]:
        containers = self.api.containers(all=all_containers)
        return [self._transform_container(container) for container in containers]

    def get_container(self, container_id: str) -> dict[str, Any]:
        details = self.api.inspect_container(container_id)
        return self._transform_container_details(details)

    def start_container(self, container_id: str) -> None:
        self.api.start(container_id)

    def stop_container(self, container_id: str) -> None:
        self.api.stop(container_id, timeout=10)

    def restart_container(self, container_id: str) -> None:
        self.api.restart(container_id, timeout=10)

    def pause_container(self, container_id: str) -> None:
        self.api.pause(container_id)

    def unpause_container(self, container_id: str) -> None:
        self.api.unpause(container_id)

    def remove_container(self, container_id: str, force: bool = False) -> None:
        self.api.remove_container(container_id, force=force, v=True)

    def get_container_stats(self, container_id: str) -> dict[str, Any]:
        stats = self.api.stats(container_id, stream=False)
        return self._transform_stats(stats, container_id)

    def open_log_stream(
        self,
        container_id: str,
        *,
        follow: bool = True,
        stdout: bool = True,
        stderr: bool = True,
        tail: str = "100",
    ) -> Iterable[bytes]:
        return self.api.logs(
            container=container_id,
            stream=True,
            follow=follow,
            stdout=stdout,
            stderr=stderr,
            tail=tail,
        )

    def exec_in_container(
        self,
        container_id: str,
        cmd: list[str],
        env: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        exec_info = self.api.exec_create(
            container=container_id,
            cmd=cmd,
            stdout=True,
            stderr=True,
            environment=[f"{key}={value}" for key, value in (env or {}).items()],
        )
        exec_id = exec_info["Id"]
        output = self.api.exec_start(exec_id, detach=False, tty=False)
        inspected = self.api.exec_inspect(exec_id)
        decoded_output = output.decode("utf-8", errors="ignore") if isinstance(output, bytes) else str(output)
        return {
            "exitCode": inspected.get("ExitCode", 0) or 0,
            "output": decoded_output,
        }

    def create_exec_socket(
        self,
        container_id: str,
        cmd: list[str],
    ) -> tuple[str, Any]:
        created = self.api.exec_create(
            container=container_id,
            cmd=cmd,
            stdin=True,
            stdout=True,
            stderr=True,
            tty=True,
        )
        exec_id = created["Id"]
        raw_socket = self.api.exec_start(
            exec_id,
            detach=False,
            tty=True,
            socket=True,
        )
        return exec_id, raw_socket

    def resize_exec(self, exec_id: str, rows: int, cols: int) -> None:
        self.api.exec_resize(exec_id, height=rows, width=cols)

    def inspect_exec(self, exec_id: str) -> dict[str, Any]:
        return self.api.exec_inspect(exec_id)

    def get_docker_version(self) -> dict[str, str]:
        version = self.api.version()
        return {
            "version": version.get("Version", "unknown"),
            "apiVersion": version.get("ApiVersion", "unknown"),
        }

    def inspect_container_state(self, container_id: str) -> dict[str, Any]:
        details = self.api.inspect_container(container_id)
        return details.get("State", {})

    def _transform_container(self, docker_container: dict[str, Any]) -> dict[str, Any]:
        return {
            "id": docker_container.get("Id", ""),
            "names": [name.lstrip("/") for name in docker_container.get("Names", [])],
            "image": docker_container.get("Image", ""),
            "imageId": docker_container.get("ImageID", ""),
            "command": docker_container.get("Command", ""),
            "created": docker_container.get("Created", 0),
            "state": docker_container.get("State", ""),
            "status": docker_container.get("Status", ""),
            "ports": [
                {
                    "IP": port.get("IP"),
                    "privatePort": port.get("PrivatePort"),
                    "publicPort": port.get("PublicPort"),
                    "type": port.get("Type", "tcp"),
                }
                for port in docker_container.get("Ports", [])
            ],
            "labels": docker_container.get("Labels") or {},
        }

    def _transform_container_details(self, details: dict[str, Any]) -> dict[str, Any]:
        container = self._transform_container(
            {
                "Id": details.get("Id", ""),
                "Names": [details.get("Name", "").lstrip("/")] if details.get("Name") else [],
                "Image": (details.get("Config") or {}).get("Image", ""),
                "ImageID": details.get("Image", ""),
                "Command": " ".join((details.get("Config") or {}).get("Cmd") or []),
                "Created": int(self._created_timestamp(details.get("Created"))),
                "State": (details.get("State") or {}).get("Status", ""),
                "Status": (details.get("State") or {}).get("Status", ""),
                "Ports": [],
                "Labels": (details.get("Config") or {}).get("Labels") or {},
            }
        )

        network_settings = details.get("NetworkSettings") or {}
        host_config = details.get("HostConfig") or {}
        config_section = details.get("Config") or {}

        return {
            **container,
            "hostConfig": {
                "portBindings": host_config.get("PortBindings"),
                "binds": host_config.get("Binds"),
                "restartPolicy": host_config.get("RestartPolicy"),
            },
            "config": {
                "labels": config_section.get("Labels"),
                "env": config_section.get("Env"),
                "cmd": config_section.get("Cmd"),
                "entrypoint": config_section.get("Entrypoint"),
                "workingDir": config_section.get("WorkingDir"),
                "user": config_section.get("User"),
            },
            "networkSettings": {
                "networks": self._transform_networks(network_settings.get("Networks") or {}),
                "ipAddress": network_settings.get("IPAddress"),
                "ipPrefixLen": network_settings.get("IPPrefixLen"),
                "gateway": network_settings.get("Gateway"),
                "bridge": network_settings.get("Bridge"),
            },
            "mounts": [
                {
                    "type": mount.get("Type", "bind"),
                    "source": mount.get("Source"),
                    "destination": mount.get("Destination", ""),
                    "mode": mount.get("Mode", ""),
                    "rw": mount.get("RW", False),
                    "propagation": mount.get("Propagation", ""),
                }
                for mount in details.get("Mounts", [])
            ],
        }

    @staticmethod
    def _transform_networks(networks: dict[str, Any]) -> dict[str, Any]:
        transformed: dict[str, Any] = {}
        for key, value in networks.items():
            transformed[key] = {
                "IPAMConfig": value.get("IPAMConfig"),
                "links": value.get("Links"),
                "aliases": value.get("Aliases"),
                "networkID": value.get("NetworkID", ""),
                "endpointID": value.get("EndpointID", ""),
                "gateway": value.get("Gateway", ""),
                "ipAddress": value.get("IPAddress", ""),
                "ipPrefixLen": value.get("IPPrefixLen", 0),
                "ipv6Gateway": value.get("IPv6Gateway", ""),
                "globalIPv6Address": value.get("GlobalIPv6Address", ""),
                "globalIPv6PrefixLen": value.get("GlobalIPv6PrefixLen", 0),
                "macAddress": value.get("MacAddress", ""),
            }
        return transformed

    @staticmethod
    def _created_timestamp(created: str | None) -> float:
        if not created:
            return 0
        created_normalized = created.replace("Z", "+00:00")
        try:
            from datetime import datetime

            return datetime.fromisoformat(created_normalized).timestamp()
        except ValueError:
            return 0

    @staticmethod
    def _transform_stats(stats: dict[str, Any], container_id: str) -> dict[str, Any]:
        cpu_stats = stats.get("cpu_stats") or {}
        precpu_stats = stats.get("precpu_stats") or {}
        cpu_usage = cpu_stats.get("cpu_usage") or {}
        precpu_usage = precpu_stats.get("cpu_usage") or {}

        cpu_delta = cpu_usage.get("total_usage", 0) - precpu_usage.get("total_usage", 0)
        system_delta = cpu_stats.get("system_cpu_usage", 0) - precpu_stats.get("system_cpu_usage", 0)
        online_cpus = cpu_stats.get("online_cpus") or 1
        cpu_percent = (cpu_delta / system_delta) * 100 * online_cpus if system_delta > 0 else 0

        memory_stats = stats.get("memory_stats") or {}
        memory_usage = memory_stats.get("usage", 0) or 0
        memory_limit = memory_stats.get("limit", 0) or 0
        memory_percent = (memory_usage / memory_limit) * 100 if memory_limit > 0 else 0

        networks = list((stats.get("networks") or {}).values())
        net_rx = sum(network.get("rx_bytes", 0) for network in networks)
        net_tx = sum(network.get("tx_bytes", 0) for network in networks)

        io_stats = ((stats.get("blkio_stats") or {}).get("io_service_bytes_recursive")) or []
        block_read = sum(io.get("value", 0) for io in io_stats if io.get("op") == "Read")
        block_write = sum(io.get("value", 0) for io in io_stats if io.get("op") == "Write")

        return {
            "name": stats.get("name", ""),
            "id": container_id,
            "cpuPercent": cpu_percent,
            "memoryUsage": memory_usage,
            "memoryLimit": memory_limit,
            "memoryPercent": memory_percent,
            "netRx": net_rx,
            "netTx": net_tx,
            "blockRead": block_read,
            "blockWrite": block_write,
        }


docker_service = DockerService()
